### PR TITLE
Fixes a bug in rad diags calculations due to prescribed volcanic emissions

### PR DESCRIPTION
**Pull request by singhbalwinder**
Radiation diagnostic calculations were producing different
answers even for the exact same list of aerosols mentioned in
the namelist in rad_diag_1 and rad_diag_2 fields. This bug
appears only for the compsets which turn on the prescribed
volcanic emissions (e.g. F2010). The bug doesn't affect other
compsets (e.g. FC5AV1C-L).
The reason was the use of unit conversion factor (km to m)
which got applied repeatedly for each diagnostic call. This bug
*does not* affect any simulation which doesn't use repeated
radiation diagnostic calls via rad_diag_1 etc. in the namelist

[BFB] - Bit-For-Bit

### CODE REVIEW

**Comment by singhbalwinder** [comment]
Thanks to Mingxuan for finding this issue.

**Comment by singhbalwinder** [comment]
Please select "hide white spaces" when reviewing the changes as I have removed the extra white spaces.

@golaz and @rljacob : This PR fixes a bug in radiation diagnostic calls, so should be included in v2, if possible.

**Comment by tangq** [comment]
It seems to me that this bug will affect the PD&PI estimation of the radiative effects for the v2 test runs.

File: components/eam/src/physics/cam/aer_rad_props.F90 [hunk 1]
```diff
@@ -231,22 +231,21 @@ subroutine aer_rad_props_sw(list_idx, dt, state, pbuf,  nnite, idxnite, is_cmip6
...
+   !Special treatment for CMIP6 volcanic aerosols, where extinction, ssa
    !and af are directly read from the prescribed volcanic aerosol file
 
-   !Point ext_cmip6_sw to unphysically huges values so that the code breaks if they are used for is_cmip6_volc=.false. case
+   !Point ext_cmip6_sw to null so that the code breaks if they are used for is_cmip6_volc=.false. case
    !This is done to avoid having optional arguments in modal_aero_sw call
-   rhuge_arr = rhuge
-   ext_cmip6_sw => rhuge_arr
-   trop_level = ihuge
+   ext_cmip6_sw => null()
```
Comment (kaizhangpnl) [inline components/eam/src/physics/cam/aer_rad_props.F90:239]: Remove rhuge as well?

**Comment by kaizhangpnl** [comment]
Hi @singhbalwinder this bug seems that all use cases with "CMIP6" (https://github.com/E3SM-Project/E3SM/tree/master/components/eam/bld/namelist_files/use_cases) are affected, right? Also, maint-1.0 needs to be fixed too.

**Comment by singhbalwinder** [comment]
Thanks Kai and Qi for your review. I have removed the `rhuge` and redefined `ihuge`. 

You are right, this fix will affect all CMIP6 "use cases" (since they all use prescribed volcanic aerosols) if we have used `rad_diag` (radiation diagnostics) namelist variable in the namelist. Simulations which didn't use `rad_diag` should remain BFB.

**Comment by singhbalwinder** [comment]
I will fix it in the maint-1.0 branch as well.

Please let me know if you would like me to run PI-PD simulations to see impact of this change.

**Comment by mingxuanwupnnl** [comment]
Thank you, Balwinder. I tested the fix, and found that the simulation would not be BFB when rad_diag_* are in the namelist. As you pointed out, we might also need to change radiation.F90 in rrtmgp/rrtmg, right?

**Review by mingxuanwupnnl — COMMENTED** [review]
The fix in aer_rad_props.F90 looks good to me. However, the model would not be BFB when rad_diag_* are in the namelist. We may also fix that.

**Comment by kaizhangpnl** [comment]
> I will fix it in the maint-1.0 branch as well.
> 
> Please let me know if you would like me to run PI-PD simulations to see impact of this change.

Yes, it would be nice to evaluate the impact. Maybe we can use your previous simulations as a reference (with the bug)? I can help with the post-processing and analysis.

**Comment by golaz** [comment]
@singhbalwinder and @rljacob : I agree, the should be merged to be in the v2 code base.

**Comment by wlin7** [comment]
> The fix in aer_rad_props.F90 looks good to me. However, the model would not be BFB when rad_diag_* are in the namelist. We may also fix that.

Hi @mingxuanwupnnl , do you mean we need additional code changes? The bug is only with the conversion factor being repeatedly multiplied to the pbuf variables, which is now fixed with new work variables. I feel like no other changes needed, but I could be wrong as I haven't run such rad_diags_* tests.

**Comment by singhbalwinder** [comment]
I also believe we do not need any additional changes.

@mingxuanwupnnl : I think even if we switch the ordering of the do-loop in radiation.F90:
`do icall = N_DIAG, 0, -1`
the results will still not be BFB with simulations which used `rad_diags_*` in the namelist in the simulations which has the bug. 

In the buggy simulations, the unit conversion multiplier will be applied at least twice (depending on how many `rad_diag_*` we have in the namelist) to the prognostic (`icall=0`) calculations. If we revert the order of the do-loop, this multiplier will be applied exactly once. Therefore the results will be non-BFB even if we change the order. Does that make sense?

I am running few simulations to access the impact of this bug fix.

**Comment by mingxuanwupnnl** [comment]
@singhbalwinder @wlin7 I'm a little confused here. Just try to check my understanding. 

When this fix is applied, it is BFB (compared with the code without fix) when rad_diag_* are not used.  

For two simulations which both apply this fix, the one that uses rad_diag_* has different values for prognostic vars (e.g., T, U, V) compared with the one that does not use rad_diag_*. Do we need to fix that? Or there are always differences for prognostic vars between with and without rad_diag_*?

**Comment by singhbalwinder** [comment]
All prognostic variables (such as T,P, U, Q, MMRs etc.) should be BFB for simulations whether we use `rad_diag_*` in the namelist or not.

Unfortunately, this bug was causing the prognostic variables to differ between the simulations run with and without rad_diag_*. It is only triggered if we turn on prescribed volcanic emission. With the bug fix in place, now the prognostic variables will be BFB in simulations run with or without rad_diag_*.

This is how I understand this code, please correct me if I got something wrong.

The reason is the way do-loop executes in radiation.F90:
`do icall = N_DIAG, 0, -1`

In a simulation with the bug, the unit conversion is applied for each iteration of this loop (which is wrong as it should be applied only once). Therefore the prognostic call (`icall=0`), which is the last one, will have unit conversion factor multiplied `N_DIAG` times. That is the reason why the prognostic variables will be different in simulation where `N_DIAG=0` (i.e. no `rad_diag_*` in the namelist) vs. `N_DIAG>0` (i.e. one or more `rad_diag_*` in the namelist).

In bug fixed simulations, unit conversion is applied only once (using a work variable instead of the `pbuf` variable) therefore the prognostic variables will not depend on the value of `N_DIAG`. 

Since using a work variable fixes this issue, we do not need to change the order of the do-loop. I haven't looked into the code but there may be other parts of the code which may be  dependent on this do-loop order. If we change the order, we might risk breaking that unless we rigorously evaluate the model after changing the order.

**Comment by wlin7** [comment]
Hi @kaizhangpnl , @mingxuanwupnnl , can you please confirm the PR is ready to go by approving the changes? Thanks.

**Review by kaizhangpnl — APPROVED** [review]

**Comment by wlin7** [comment]
Merged to next.

**Review by mingxuanwupnnl — APPROVED** [review]

**Comment by wlin7** [comment]
Merged to master.

**Pull request merged**